File: mRemoteV1/Config/Putty/XmingProvider.vb
Imports System.IO
Imports System.Text.RegularExpressions
Imports System.Web
Imports mRemote3G.App
Imports mRemote3G.App.Info
Imports mRemote3G.Connection.Protocol
Imports mRemote3G.Connection.PuttySession
Imports mRemote3G.Messages

Namespace Config.Putty
    Public Class XmingProvider
        Inherits Provider

#Region "Public Methods"

        Public Overrides Function GetSessionNames(Optional ByVal raw As Boolean = False) As String()
            Dim sessionsFolderPath As String = GetSessionsFolderPath()
            If Not Directory.Exists(sessionsFolderPath) Then Return New String() {}

            Dim sessionNames As New List(Of String)
            For Each sessionName As String In Directory.GetFiles(sessionsFolderPath)
                sessionName = Path.GetFileName(sessionName)
                If raw Then
                    sessionNames.Add(sessionName)
                Else
                    sessionNames.Add(HttpUtility.UrlDecode(sessionName.Replace("+", "%2B")))
                End If
            Next

            If raw Then
                If Not sessionNames.Contains("Default%20Settings") Then ' Do not localize
                    sessionNames.Insert(0, "Default%20Settings")
                End If
            Else
                If Not sessionNames.Contains("Default Settings") Then
                    sessionNames.Insert(0, "Default Settings")
                End If
            End If

            Dim registrySessionNames As New List(Of String)
            For Each sessionName As String In RegistryProvider.GetSessionNames(raw)
                registrySessionNames.Add(String.Format(RegistrySessionNameFormat, sessionName))
            Next

            sessionNames.AddRange(registrySessionNames)
            sessionNames.Sort()

            Return sessionNames.ToArray()
        End Function

        Public Overrides Function GetSession(sessionName As String) As PuttyInfo
            Dim registrySessionName As String = GetRegistrySessionName(sessionName)
            If Not String.IsNullOrEmpty(registrySessionName) Then
                Return ModifyRegistrySessionInfo(RegistryProvider.GetSession(registrySessionName))
            End If

            Dim sessionsFolderPath As String = GetSessionsFolderPath()
            If Not Directory.Exists(sessionsFolderPath) Then Return Nothing

            Dim sessionFile As String = Path.Combine(sessionsFolderPath, sessionName)
            If Not File.Exists(sessionFile) Then Return Nothing

            sessionName = HttpUtility.UrlDecode(sessionName.Replace("+", "%2B"))

            Dim sessionFileReader As New SessionFileReader(sessionFile)
            Dim sessionInfo As New PuttyInfo
            With sessionInfo
                .PuttySession = sessionName
                .Name = sessionName
                .Hostname = sessionFileReader.GetValue("HostName")
                .Username = sessionFileReader.GetValue("UserName")
                Dim protocol As String = sessionFileReader.GetValue("Protocol")
                If protocol Is Nothing Then protocol = "ssh"
                Select Case protocol.ToLowerInvariant()
                    Case "raw"
                        .Protocol = Protocols.RAW
                    Case "rlogin"
                        .Protocol = Protocols.Rlogin
                    Case "serial"
                        Return Nothing
                    Case "ssh"
                        Dim sshVersionObject As Object = sessionFileReader.GetValue("SshProt")
                        If sshVersionObject IsNot Nothing Then
                            Dim sshVersion = CType(sshVersionObject, Integer)
                            If sshVersion >= 2 Then
                                .Protocol = Protocols.SSH2
                            Else
                                .Protocol = Protocols.SSH1
                            End If
                        Else
                            .Protocol = Protocols.SSH2
                        End If
                    Case "telnet"
                        .Protocol = Protocols.Telnet
                    Case Else
                        Return Nothing
                End Select
                .Port = sessionFileReader.GetValue("PortNumber")
            End With

            Return sessionInfo
        End Function

        Public Overrides Sub StartWatcher()
            RegistryProvider.StartWatcher()
            AddHandler RegistryProvider.SessionChanged, AddressOf OnRegistrySessionChanged

            If _eventWatcher IsNot Nothing Then Return

            Dim sessDir = GetSessionsFolderPath()
            If Not Directory.Exists(sessDir) Then
                Runtime.MessageCollector.AddMessage(MessageClass.InformationMsg, "XmingProvider: Could not find Session Dir. Cannot watch sessions.", True)
                Return
            End If

            Try
                _eventWatcher = New FileSystemWatcher(sessDir)
                _eventWatcher.NotifyFilter = (NotifyFilters.FileName Or NotifyFilters.LastWrite)
                AddHandler _eventWatcher.Changed, AddressOf OnFileSystemEventArrived
                AddHandler _eventWatcher.Created, AddressOf OnFileSystemEventArrived
                AddHandler _eventWatcher.Deleted, AddressOf OnFileSystemEventArrived
                AddHandler _eventWatcher.Renamed, AddressOf OnFileSystemEventArrived
                _eventWatcher.EnableRaisingEvents = True
            Catch ex As Exception
                Runtime.MessageCollector.AddExceptionMessage(
                    "XmingPortablePuttySessions.Watcher.StartWatching() failed.", ex, MessageClass.WarningMsg, True)
            End Try
        End Sub

        Public Overrides Sub StopWatcher()
            RegistryProvider.StopWatcher()
            RemoveHandler RegistryProvider.SessionChanged, AddressOf OnRegistrySessionChanged

            If _eventWatcher Is Nothing Then Return
            _eventWatcher.EnableRaisingEvents = False
            _eventWatcher.Dispose()
            _eventWatcher = Nothing
        End Sub

#End Region

#Region "Private Fields"

        Private Const RegistrySessionNameFormat As String = "{0} [registry]"
        Private Const RegistrySessionNamePattern As String = "(.*)\ \[registry\]"

        Private Shared ReadOnly RegistryProvider As New RegistryProvider
        Private Shared _eventWatcher As FileSystemWatcher

#End Region

#Region "Private Methods"

        Private Shared Function GetPuttyConfPath() As String
            Dim puttyPath As String
            If My.Settings.UseCustomPuttyPath Then
                puttyPath = My.Settings.CustomPuttyPath
            Else
                puttyPath = General.PuttyPath
            End If
            Return Path.Combine(Path.GetDirectoryName(puttyPath), "putty.conf")
        End Function

        Private Shared Function GetSessionsFolderPath() As String
            Dim puttyConfPath As String = GetPuttyConfPath()
            Dim sessionFileReader As New PuttyConfFileReader(puttyConfPath)
            Dim basePath As String = Environment.ExpandEnvironmentVariables(sessionFileReader.GetValue("sshk&sess"))
            Return Path.Combine(basePath, "sessions")
        End Function

        Private Shared Function GetRegistrySessionName(sessionName As String) As String
            Dim regex As New Regex(RegistrySessionNamePattern)

            Dim matches As MatchCollection = regex.Matches(sessionName)
            If matches.Count < 1 Then Return String.Empty

            Dim groups As GroupCollection = matches(0).Groups
            If groups.Count < 1 Then Return String.Empty _
            ' This should always include at least one item, but check anyway

            Return groups(1).Value
        End Function

        Private Shared Function ModifyRegistrySessionInfo(sessionInfo As PuttyInfo) As PuttyInfo
            sessionInfo.Name = String.Format(RegistrySessionNameFormat, sessionInfo.Name)
            sessionInfo.PuttySession = String.Format(RegistrySessionNameFormat, sessionInfo.PuttySession)
            Return sessionInfo
        End Function

        Private Sub OnFileSystemEventArrived(sender As Object, e As FileSystemEventArgs)
            OnSessionChanged(New SessionChangedEventArgs())
        End Sub

        Private Sub OnRegistrySessionChanged(sender As Object, e As SessionChangedEventArgs)
            OnSessionChanged(New SessionChangedEventArgs())
        End Sub

#End Region

#Region "Private Classes"

        Private Class PuttyConfFileReader
            Public Sub New(puttyConfFile As String)
                _puttyConfFile = puttyConfFile
            End Sub

            Private ReadOnly _puttyConfFile As String
            Private _configurationLoaded As Boolean = False
            Private ReadOnly _configuration As New Dictionary(Of String, String)

            Private Sub LoadConfiguration()
                _configurationLoaded = True
                Try
                    If Not File.Exists(_puttyConfFile) Then Return
                    Using streamReader As New StreamReader(_puttyConfFile)
                        Dim line As String
                        Do
                            line = streamReader.ReadLine()
                            If line Is Nothing Then Exit Do
                            line = line.Trim()
                            If line = String.Empty Then Continue Do ' Blank line
                            If line.Substring(0, 1) = ";" Then Continue Do ' Comment
                            Dim parts() As String = line.Split(New Char() {"="}, 2)
                            If parts.Length < 2 Then Continue Do
                            If _configuration.ContainsKey(parts(0)) Then Continue Do _
                            ' As per http://www.straightrunning.com/XmingNotes/portableputty.php only first entry is used
                            _configuration.Add(parts(0), parts(1))
                        Loop
                    End Using
                Catch ex As Exception
                    Runtime.MessageCollector.AddExceptionMessage("PuttyConfFileReader.LoadConfiguration() failed.", ex,
                                                                 MessageClass.ErrorMsg, True)
                End Try
            End Sub

            Public Function GetValue(setting As String) As String
                If Not _configurationLoaded Then LoadConfiguration()
                If Not _configuration.ContainsKey(setting) Then Return String.Empty
                Return _configuration(setting)
            End Function
        End Class

        Private Class SessionFileReader
            Public Sub New(sessionFile As String)
                _sessionFile = sessionFile
            End Sub

            Private ReadOnly _sessionFile As String
            Private _sessionInfoLoaded As Boolean = False
            Private ReadOnly _sessionInfo As New Dictionary(Of String, String)

            Private Sub LoadSessionInfo()
                _sessionInfoLoaded = True
                Try
                    If Not File.Exists(_sessionFile) Then Return
                    Using streamReader As New StreamReader(_sessionFile)
                        Dim line As String
                        Do
                            line = streamReader.ReadLine()
                            If line Is Nothing Then Exit Do
                            Dim parts() As String = line.Split(New Char() {"\"})
                            If parts.Length < 2 Then Continue Do
                            _sessionInfo.Add(parts(0), parts(1))
                        Loop
                    End Using
                Catch ex As Exception
                    Runtime.MessageCollector.AddExceptionMessage("SessionFileReader.LoadSessionInfo() failed.", ex,
                                                                 MessageClass.ErrorMsg, True)
                End Try
            End Sub

            Public Function GetValue(setting As String) As String
                If Not _sessionInfoLoaded Then LoadSessionInfo()
                If Not _sessionInfo.ContainsKey(setting) Then Return String.Empty
                Return _sessionInfo(setting)
            End Function
        End Class

#End Region
    End Class
End Namespace

File: mRemoteV1.ConvertedToC#/Config/Putty/XmingProvider.cs
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Drawing;
using System.Diagnostics;
using System.Windows.Forms;
using System.IO;
using mRemoteNG.App;
using mRemoteNG.Messages;
using mRemoteNG.Connection.Protocol;
using System.Text.RegularExpressions;

namespace mRemoteNG.Config.Putty
{
	public class XmingProvider : Provider
	{

		#region "Public Methods"
		public override string[] GetSessionNames(bool raw = false)
		{
			string sessionsFolderPath = GetSessionsFolderPath();
			if (!Directory.Exists(sessionsFolderPath))
				return new string[];

			List<string> sessionNames = new List<string>();
			foreach (string sessionName in Directory.GetFiles(sessionsFolderPath)) {
				sessionName = Path.GetFileName(sessionName);
				if (raw) {
					sessionNames.Add(sessionName);
				} else {
					sessionNames.Add(System.Web.HttpUtility.UrlDecode(sessionName.Replace("+", "%2B")));
				}
			}

			if (raw) {
				// Do not localize
				if (!sessionNames.Contains("Default%20Settings")) {
					sessionNames.Insert(0, "Default%20Settings");
				}
			} else {
				if (!sessionNames.Contains("Default Settings")) {
					sessionNames.Insert(0, "Default Settings");
				}
			}

			List<string> registrySessionNames = new List<string>();
			foreach (string sessionName in RegistryProvider.GetSessionNames(raw)) {
				registrySessionNames.Add(string.Format(RegistrySessionNameFormat, sessionName));
			}

			sessionNames.AddRange(registrySessionNames);
			sessionNames.Sort();

			return sessionNames.ToArray();
		}

		public override Connection.PuttySession.Info GetSession(string sessionName)
		{
			string registrySessionName = GetRegistrySessionName(sessionName);
			if (!string.IsNullOrEmpty(registrySessionName)) {
				return ModifyRegistrySessionInfo(RegistryProvider.GetSession(registrySessionName));
			}

			string sessionsFolderPath = GetSessionsFolderPath();
			if (!Directory.Exists(sessionsFolderPath))
				return null;

			string sessionFile = Path.Combine(sessionsFolderPath, sessionName);
			if (!File.Exists(sessionFile))
				return null;

			sessionName = System.Web.HttpUtility.UrlDecode(sessionName.Replace("+", "%2B"));

			SessionFileReader sessionFileReader = new SessionFileReader(sessionFile);
			Connection.PuttySession.Info sessionInfo = new Connection.PuttySession.Info();
			var _with1 = sessionInfo;
			_with1.PuttySession = sessionName;
			_with1.Name = sessionName;
			_with1.Hostname = sessionFileReader.GetValue("HostName");
			_with1.Username = sessionFileReader.GetValue("UserName");
			string protocol = sessionFileReader.GetValue("Protocol");
			if (protocol == null)
				protocol = "ssh";
			switch (protocol.ToLowerInvariant()) {
				case "raw":
					_with1.Protocol = Protocols.RAW;
					break;
				case "rlogin":
					_with1.Protocol = Protocols.Rlogin;
					break;
				case "serial":
					return null;
				case "ssh":
					object sshVersionObject = sessionFileReader.GetValue("SshProt");
					if (sshVersionObject != null) {
						int sshVersion = Convert.ToInt32(sshVersionObject);
						if (sshVersion >= 2) {
							_with1.Protocol = Protocols.SSH2;
						} else {
							_with1.Protocol = Protocols.SSH1;
						}
					} else {
						_with1.Protocol = Protocols.SSH2;
					}
					break;
				case "telnet":
					_with1.Protocol = Protocols.Telnet;
					break;
				default:
					return null;
			}
			_with1.Port = sessionFileReader.GetValue("PortNumber");

			return sessionInfo;
		}

		public override void StartWatcher()
		{
			RegistryProvider.StartWatcher();
			RegistryProvider.SessionChanged += OnRegistrySessionChanged;

			if (_eventWatcher != null)
				return;

			try {
				_eventWatcher = new FileSystemWatcher(GetSessionsFolderPath());
				_eventWatcher.NotifyFilter = (NotifyFilters.FileName | NotifyFilters.LastWrite);
				_eventWatcher.Changed += OnFileSystemEventArrived;
				_eventWatcher.Created += OnFileSystemEventArrived;
				_eventWatcher.Deleted += OnFileSystemEventArrived;
				_eventWatcher.Renamed += OnFileSystemEventArrived;
				_eventWatcher.EnableRaisingEvents = true;
			} catch (Exception ex) {
				Runtime.MessageCollector.AddExceptionMessage("XmingPortablePuttySessions.Watcher.StartWatching() failed.", ex, MessageClass.WarningMsg, true);
			}
		}

		public override void StopWatcher()
		{
			RegistryProvider.StopWatcher();
			RegistryProvider.SessionChanged -= OnRegistrySessionChanged;

			if (_eventWatcher == null)
				return;
			_eventWatcher.EnableRaisingEvents = false;
			_eventWatcher.Dispose();
			_eventWatcher = null;
		}
		#endregion

		#region "Private Fields"
		private const string RegistrySessionNameFormat = "{0} [registry]";

		private const string RegistrySessionNamePattern = "(.*)\\ \\[registry\\]";
		private static readonly RegistryProvider RegistryProvider = new RegistryProvider();
			#endregion
		private static FileSystemWatcher _eventWatcher;

		#region "Private Methods"
		private static string GetPuttyConfPath()
		{
			string puttyPath = null;
			if (mRemoteNG.My.Settings.UseCustomPuttyPath) {
				puttyPath = mRemoteNG.My.Settings.CustomPuttyPath;
			} else {
				puttyPath = mRemoteNG.App.Info.General.PuttyPath;
			}
			return Path.Combine(Path.GetDirectoryName(puttyPath), "putty.conf");
		}

		private static string GetSessionsFolderPath()
		{
			string puttyConfPath = GetPuttyConfPath();
			PuttyConfFileReader sessionFileReader = new PuttyConfFileReader(puttyConfPath);
			string basePath = Environment.ExpandEnvironmentVariables(sessionFileReader.GetValue("sshk&sess"));
			return Path.Combine(basePath, "sessions");
		}

		private static string GetRegistrySessionName(string sessionName)
		{
			Regex regex = new Regex(RegistrySessionNamePattern);

			MatchCollection matches = regex.Matches(sessionName);
			if (matches.Count < 1)
				return string.Empty;

			GroupCollection groups = matches[0].Groups;
			if (groups.Count < 1)
				return string.Empty;
			// This should always include at least one item, but check anyway

			return groups[1].Value;
		}

		private static Connection.PuttySession.Info ModifyRegistrySessionInfo(Connection.PuttySession.Info sessionInfo)
		{
			sessionInfo.Name = string.Format(RegistrySessionNameFormat, sessionInfo.Name);
			sessionInfo.PuttySession = string.Format(RegistrySessionNameFormat, sessionInfo.PuttySession);
			return sessionInfo;
		}

		private void OnFileSystemEventArrived(object sender, FileSystemEventArgs e)
		{
			OnSessionChanged(new SessionChangedEventArgs());
		}

		private void OnRegistrySessionChanged(object sender, SessionChangedEventArgs e)
		{
			OnSessionChanged(new SessionChangedEventArgs());
		}
		#endregion

		#region "Private Classes"
		private class PuttyConfFileReader
		{
			public PuttyConfFileReader(string puttyConfFile)
			{
				_puttyConfFile = puttyConfFile;
			}

			private readonly string _puttyConfFile;
			private bool _configurationLoaded = false;

			private readonly Dictionary<string, string> _configuration = new Dictionary<string, string>();
			private void LoadConfiguration()
			{
				_configurationLoaded = true;
				try {
					if (!File.Exists(_puttyConfFile))
						return;
					using (StreamReader streamReader = new StreamReader(_puttyConfFile)) {
						string line = null;
						do {
							line = streamReader.ReadLine();
							if (line == null)
								break; // TODO: might not be correct. Was : Exit Do
							line = line.Trim();
							if (line == string.Empty)
								continue;
							// Blank line
							if (line.Substring(0, 1) == ";")
								continue;
							// Comment
							string[] parts = line.Split(new char[] { "=" }, 2);
							if (parts.Length < 2)
								continue;
							if (_configuration.ContainsKey(parts[0]))
								continue;
							// As per http://www.straightrunning.com/XmingNotes/portableputty.php only first entry is used
							_configuration.Add(parts[0], parts[1]);
						} while (true);
					}
				} catch (Exception ex) {
					Runtime.MessageCollector.AddExceptionMessage("PuttyConfFileReader.LoadConfiguration() failed.", ex, MessageClass.ErrorMsg, true);
				}
			}

			public string GetValue(string setting)
			{
				if (!_configurationLoaded)
					LoadConfiguration();
				if (!_configuration.ContainsKey(setting))
					return string.Empty;
				return _configuration[setting];
			}
		}

		private class SessionFileReader
		{
			public SessionFileReader(string sessionFile)
			{
				_sessionFile = sessionFile;
			}

			private readonly string _sessionFile;
			private bool _sessionInfoLoaded = false;

			private readonly Dictionary<string, string> _sessionInfo = new Dictionary<string, string>();
			private void LoadSessionInfo()
			{
				_sessionInfoLoaded = true;
				try {
					if (!File.Exists(_sessionFile))
						return;
					using (StreamReader streamReader = new StreamReader(_sessionFile)) {
						string line = null;
						do {
							line = streamReader.ReadLine();
							if (line == null)
								break; // TODO: might not be correct. Was : Exit Do
							string[] parts = line.Split(new char[] { "\\" });
							if (parts.Length < 2)
								continue;
							_sessionInfo.Add(parts[0], parts[1]);
						} while (true);
					}
				} catch (Exception ex) {
					Runtime.MessageCollector.AddExceptionMessage("SessionFileReader.LoadSessionInfo() failed.", ex, MessageClass.ErrorMsg, true);
				}
			}

			public string GetValue(string setting)
			{
				if (!_sessionInfoLoaded)
					LoadSessionInfo();
				if (!_sessionInfo.ContainsKey(setting))
					return string.Empty;
				return _sessionInfo[setting];
			}
		}
		#endregion
	}
}
